To run this, press "*Runtime*" and press "*Run all*" on a **free** Tesla T4 Google Colab instance!
<div class="align-center">
<a href="https://unsloth.ai/"><img src="https://github.com/unslothai/unsloth/raw/main/images/unsloth%20new%20logo.png" width="115"></a>
<a href="https://discord.gg/unsloth"><img src="https://github.com/unslothai/unsloth/raw/main/images/Discord button.png" width="145"></a>
<a href="https://docs.unsloth.ai/"><img src="https://github.com/unslothai/unsloth/blob/main/images/documentation%20green%20button.png?raw=true" width="125"></a></a> Join Discord if you need help + ⭐ <i>Star us on <a href="https://github.com/unslothai/unsloth">Github</a> </i> ⭐
</div>

To install Unsloth on your own computer, follow the installation instructions on our Github page [here](https://docs.unsloth.ai/get-started/installing-+-updating).

You will learn how to do [data prep](#Data), how to [train](#Train), how to [run the model](#Inference), & [how to save it](#Save)


### News

Unsloth now supports Text-to-Speech (TTS) models. Read our [guide here](https://docs.unsloth.ai/basics/text-to-speech-tts-fine-tuning).

Read our **[Qwen3 Guide](https://docs.unsloth.ai/basics/qwen3-how-to-run-and-fine-tune)** and check out our new **[Dynamic 2.0](https://docs.unsloth.ai/basics/unsloth-dynamic-2.0-ggufs)** quants which outperforms other quantization methods!

Visit our docs for all our [model uploads](https://docs.unsloth.ai/get-started/all-our-models) and [notebooks](https://docs.unsloth.ai/get-started/unsloth-notebooks).


### Installation

In [1]:
%%capture
import os
if "COLAB_" not in "".join(os.environ.keys()):
    !pip install unsloth
else:
    # Do this only in Colab notebooks! Otherwise use pip install unsloth
    !pip install --no-deps bitsandbytes accelerate xformers==0.0.29.post3 peft trl triton cut_cross_entropy unsloth_zoo
    !pip install sentencepiece protobuf "datasets>=3.4.1" huggingface_hub hf_transfer
    !pip install --no-deps unsloth
    !pip install "unsloth[colab-new]==2025.5.7"
    !pip install "unsloth-zoo==2025.5.8"
    !pip install "transformers==4.51.3"

### Unsloth

`FastModel` supports loading nearly any model now! This includes Vision and Text models!

In [2]:
from unsloth import FastModel
import torch

fourbit_models = [
    # 4bit dynamic quants for superior accuracy and low memory use
    "unsloth/gemma-3-1b-it-unsloth-bnb-4bit",
    "unsloth/gemma-3-4b-it-unsloth-bnb-4bit",
    "unsloth/gemma-3-12b-it-unsloth-bnb-4bit",
    "unsloth/gemma-3-27b-it-unsloth-bnb-4bit",

] # More models at https://huggingface.co/unsloth

model, tokenizer = FastModel.from_pretrained(
    model_name = "unsloth/gemma-3-4b-it-qat",
    max_seq_length = 2048, # Choose any for long context!
    load_in_4bit = True,  # 4 bit quantization to reduce memory
    load_in_8bit = False, # [NEW!] A bit more accurate, uses 2x memory
    full_finetuning = False, # [NEW!] We have full finetuning now!
    # token = "hf_...", # use one if using gated models
)

🦥 Unsloth: Will patch your computer to enable 2x faster free finetuning.
🦥 Unsloth Zoo will now patch everything to make training faster!
==((====))==  Unsloth 2025.5.7: Fast Gemma3 patching. Transformers: 4.51.3.
   \\   /|    NVIDIA A100-SXM4-40GB. Num GPUs = 1. Max memory: 39.557 GB. Platform: Linux.
O^O/ \_/ \    Torch: 2.6.0+cu124. CUDA: 8.0. CUDA Toolkit: 12.4. Triton: 3.2.0
\        /    Bfloat16 = TRUE. FA [Xformers = 0.0.29.post3. FA2 = False]
 "-____-"     Free license: http://github.com/unslothai/unsloth
Unsloth: Fast downloading is enabled - ignore downloading bars which are red colored!


model.safetensors.index.json:   0%|          | 0.00/90.6k [00:00<?, ?B/s]

model-00001-of-00002.safetensors:   0%|          | 0.00/4.96G [00:00<?, ?B/s]

model-00002-of-00002.safetensors:   0%|          | 0.00/3.64G [00:00<?, ?B/s]

Loading checkpoint shards:   0%|          | 0/2 [00:00<?, ?it/s]

generation_config.json:   0%|          | 0.00/210 [00:00<?, ?B/s]

processor_config.json:   0%|          | 0.00/70.0 [00:00<?, ?B/s]

chat_template.json:   0%|          | 0.00/1.61k [00:00<?, ?B/s]

chat_template.jinja:   0%|          | 0.00/1.53k [00:00<?, ?B/s]

preprocessor_config.json:   0%|          | 0.00/570 [00:00<?, ?B/s]

Using a slow image processor as `use_fast` is unset and a slow processor was saved with this model. `use_fast=True` will be the default behavior in v4.52, even if the model was saved with a slow processor. This will result in minor differences in outputs. You'll still be able to use a slow processor with `use_fast=False`.


tokenizer_config.json:   0%|          | 0.00/1.16M [00:00<?, ?B/s]

tokenizer.model:   0%|          | 0.00/4.69M [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/33.4M [00:00<?, ?B/s]

added_tokens.json:   0%|          | 0.00/35.0 [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/670 [00:00<?, ?B/s]

We now add LoRA adapters so we only need to update a small amount of parameters!

In [3]:
model = FastModel.get_peft_model(
    model,
    finetune_vision_layers     = False, # Turn off for just text!
    finetune_language_layers   = True,  # Should leave on!
    finetune_attention_modules = True,  # Attention good for GRPO
    finetune_mlp_modules       = True,  # SHould leave on always!

    r = 8,           # Larger = higher accuracy, but might overfit
    lora_alpha = 8,  # Recommended alpha == r at least
    lora_dropout = 0,
    bias = "none",
    random_state = 3407,
)

Unsloth: Making `model.base_model.model.language_model.model` require gradients


<a name="Data"></a>
### Data Prep
We now use the `Gemma-3` format for conversation style finetunes. We use [Maxime Labonne's FineTome-100k](https://huggingface.co/datasets/mlabonne/FineTome-100k) dataset in ShareGPT style. Gemma-3 renders multi turn conversations like below:

```
<bos><start_of_turn>user
Hello!<end_of_turn>
<start_of_turn>model
Hey there!<end_of_turn>
```

We use our `get_chat_template` function to get the correct chat template. We support `zephyr, chatml, mistral, llama, alpaca, vicuna, vicuna_old, phi3, llama3, phi4, qwen2.5, gemma3` and more.

In [4]:
from unsloth.chat_templates import get_chat_template
tokenizer = get_chat_template(
    tokenizer,
    chat_template = "gemma-3",
)

In [23]:
from datasets import load_dataset
dataset = load_dataset("andresgr96/bt_dataset_a_v1", split = "train")

README.md:   0%|          | 0.00/349 [00:00<?, ?B/s]

train-00000-of-00001.parquet:   0%|          | 0.00/22.2k [00:00<?, ?B/s]

Generating train split:   0%|          | 0/300 [00:00<?, ? examples/s]

We now use `standardize_data_formats` to try converting datasets to the correct format for finetuning purposes!

In [24]:
from unsloth.chat_templates import standardize_data_formats
dataset = standardize_data_formats(dataset)

Unsloth: Standardizing formats (num_proc=12):   0%|          | 0/300 [00:00<?, ? examples/s]

Let's see how row 100 looks like!

In [25]:
dataset[100]

{'conversations': [{'content': '    You are one of the top robotics professionals in the world. You specialize in using behavior trees in XML format to control agents and robots. You are working on a project that is \n    harnessing LLMs for the control of agents in specific environments by allowing a user to specify a task in natural language, and the LLM will produce a behavior tree that should solve that task. \n    The project is focused on an environment regarding base maintenance through a swarm of agents.\n\n    Environment Description:  In the environment, several parts are scattered around and parts have two types: good parts and scrap parts. The environment consists of 3 main areas. \n    The first zone is the base where agents spawn and is separated between two smaller areas, a storage area and a construction area. The second area is the source area, good parts spawn much more frequently \n    the closer you get to the source, and bad parts while they can be found there, the

We now have to apply the chat template for `Gemma-3` onto the conversations, and save it to `text`. We remove the `<bos>` token using removeprefix(`'<bos>'`) since we're finetuning. The Processor will add this token before training and the model expects only one.

In [9]:
def formatting_prompts_func(examples):
   convos = examples["conversations"]
   texts = [tokenizer.apply_chat_template(convo, tokenize = False, add_generation_prompt = False).removeprefix('<bos>') for convo in convos]
   return { "text" : texts, }

dataset = dataset.map(formatting_prompts_func, batched = True)

Map:   0%|          | 0/300 [00:00<?, ? examples/s]

Let's see how the chat template did! Notice there is no `<bos>` token as the processor tokenizer will be adding one.

In [10]:
dataset[100]["text"]

'<start_of_turn>user\n    You are one of the top robotics professionals in the world. You specialize in using behavior trees in XML format to control agents and robots. You are working on a project that is \n    harnessing LLMs for the control of agents in specific environments by allowing a user to specify a task in natural language, and the LLM will produce a behavior tree that should solve that task. \n    The project is focused on an environment regarding base maintenance through a swarm of agents.\n\n    Environment Description:  In the environment, several parts are scattered around and parts have two types: good parts and scrap parts. The environment consists of 3 main areas. \n    The first zone is the base where agents spawn and is separated between two smaller areas, a storage area and a construction area. The second area is the source area, good parts spawn much more frequently \n    the closer you get to the source, and bad parts while they can be found there, they can appe

<a name="Train"></a>
### Train the model
Now let's use Huggingface TRL's `SFTTrainer`! More docs here: [TRL SFT docs](https://huggingface.co/docs/trl/sft_trainer). We do 60 steps to speed things up, but you can set `num_train_epochs=1` for a full run, and turn off `max_steps=None`.

In [11]:
from trl import SFTTrainer, SFTConfig
trainer = SFTTrainer(
    model = model,
    tokenizer = tokenizer,
    train_dataset = dataset,
    eval_dataset = None, # Can set up evaluation!
    args = SFTConfig(
        dataset_text_field = "text",
        per_device_train_batch_size = 2,
        gradient_accumulation_steps = 4, # Use GA to mimic batch size!
        warmup_steps = 5,
        # num_train_epochs = 1, # Set this for 1 full training run.
        max_steps = 30,
        learning_rate = 2e-4, # Reduce to 2e-5 for long training runs
        logging_steps = 1,
        optim = "adamw_8bit",
        weight_decay = 0.01,
        lr_scheduler_type = "linear",
        seed = 3407,
        report_to = "none", # Use this for WandB etc
        dataset_num_proc=2,
    ),
)

Unsloth: Tokenizing ["text"] (num_proc=2):   0%|          | 0/300 [00:00<?, ? examples/s]

We also use Unsloth's `train_on_completions` method to only train on the assistant outputs and ignore the loss on the user's inputs. This helps increase accuracy of finetunes!

In [12]:
from unsloth.chat_templates import train_on_responses_only
trainer = train_on_responses_only(
    trainer,
    instruction_part = "<start_of_turn>user\n",
    response_part = "<start_of_turn>model\n",
)

Map (num_proc=12):   0%|          | 0/300 [00:00<?, ? examples/s]

Let's verify masking the instruction part is done! Let's print the 100th row again.  Notice how the sample only has a single `<bos>` as expected!

In [13]:
tokenizer.decode(trainer.train_dataset[100]["input_ids"])

'<bos><start_of_turn>user\n    You are one of the top robotics professionals in the world. You specialize in using behavior trees in XML format to control agents and robots. You are working on a project that is \n    harnessing LLMs for the control of agents in specific environments by allowing a user to specify a task in natural language, and the LLM will produce a behavior tree that should solve that task. \n    The project is focused on an environment regarding base maintenance through a swarm of agents.\n\n    Environment Description:  In the environment, several parts are scattered around and parts have two types: good parts and scrap parts. The environment consists of 3 main areas. \n    The first zone is the base where agents spawn and is separated between two smaller areas, a storage area and a construction area. The second area is the source area, good parts spawn much more frequently \n    the closer you get to the source, and bad parts while they can be found there, they can

Now let's print the masked out example - you should see only the answer is present:

In [14]:
tokenizer.decode([tokenizer.pad_token_id if x == -100 else x for x in trainer.train_dataset[100]["labels"]]).replace(tokenizer.pad_token, " ")

'                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                       

In [15]:
# @title Show current memory stats
gpu_stats = torch.cuda.get_device_properties(0)
start_gpu_memory = round(torch.cuda.max_memory_reserved() / 1024 / 1024 / 1024, 3)
max_memory = round(gpu_stats.total_memory / 1024 / 1024 / 1024, 3)
print(f"GPU = {gpu_stats.name}. Max memory = {max_memory} GB.")
print(f"{start_gpu_memory} GB of memory reserved.")

GPU = NVIDIA A100-SXM4-40GB. Max memory = 39.557 GB.
4.891 GB of memory reserved.


Let's train the model! To resume a training run, set `trainer.train(resume_from_checkpoint = True)`

In [16]:
trainer_stats = trainer.train()

==((====))==  Unsloth - 2x faster free finetuning | Num GPUs used = 1
   \\   /|    Num examples = 300 | Num Epochs = 1 | Total steps = 30
O^O/ \_/ \    Batch size per device = 2 | Gradient accumulation steps = 4
\        /    Data Parallel GPUs = 1 | Total batch size (2 x 4 x 1) = 8
 "-____-"     Trainable parameters = 14,901,248/4,000,000,000 (0.37% trained)
`use_cache=True` is incompatible with gradient checkpointing. Setting `use_cache=False`.


Unsloth: Will smartly offload gradients to save VRAM!


In [17]:
# @title Show final memory and time stats
used_memory = round(torch.cuda.max_memory_reserved() / 1024 / 1024 / 1024, 3)
used_memory_for_lora = round(used_memory - start_gpu_memory, 3)
used_percentage = round(used_memory / max_memory * 100, 3)
lora_percentage = round(used_memory_for_lora / max_memory * 100, 3)
print(f"{trainer_stats.metrics['train_runtime']} seconds used for training.")
print(
    f"{round(trainer_stats.metrics['train_runtime']/60, 2)} minutes used for training."
)
print(f"Peak reserved memory = {used_memory} GB.")
print(f"Peak reserved memory for training = {used_memory_for_lora} GB.")
print(f"Peak reserved memory % of max memory = {used_percentage} %.")
print(f"Peak reserved memory for training % of max memory = {lora_percentage} %.")

187.7858 seconds used for training.
3.13 minutes used for training.
Peak reserved memory = 5.184 GB.
Peak reserved memory for training = 0.293 GB.
Peak reserved memory % of max memory = 13.105 %.
Peak reserved memory for training % of max memory = 0.741 %.


<a name="Inference"></a>
### Inference
Let's run the model via Unsloth native inference! According to the `Gemma-3` team, the recommended settings for inference are `temperature = 1.0, top_p = 0.95, top_k = 64`

In [22]:
from unsloth.chat_templates import get_chat_template

# Apply chat template
tokenizer = get_chat_template(
    tokenizer,
    chat_template="gemma-3",
)

# Define messages with user prompt formatted cleanly
messages = [{
    "role": "user",
    "content": [{
        "type": "text",
        "text": (
            "You are one of the top robotics professionals in the world. You specialize in using behavior trees in XML format to control agents and robots. "
            "You are working on a project that is harnessing LLMs for the control of agents in specific environments by allowing a user to specify a task in natural language, "
            "and the LLM will produce a behavior tree that should solve that task. The project is focused on an environment regarding base maintenance through a swarm of agents.\n\n"
            "Environment Description: In the environment, several parts are scattered around and parts have two types: good parts and scrap parts. The environment consists of 3 main areas. "
            "The first zone is the base where agents spawn and is separated between two smaller areas, a storage area and a construction area. The second area is the source area; "
            "good parts spawn much more frequently the closer you get to the source, and bad parts while they can be found there, they can appear anywhere except the base area. "
            "The last area is the waste area, where agents can bring and drop scrap parts so they are put away so they don't interfere with other agents.\n\n"
            "Your task is to generate well-structured XML code for behavior trees based on the provided instructions. Only output the raw XML code, no other text and don't add markdown code tags. "
            "Use only the following types of node types:\n\n"
            "- BehaviorTree: Always the root node, and can only have one child (Sequence or Selector).\n"
            "- Sequence: Executes children top-to-bottom; halts if any child fails.\n"
            "- Selector: Executes children top-to-bottom; succeeds on first success.\n"
            "- Condition: Checks environment or internal robot state.\n"
            "- ActuatorAction: Controls actuators (e.g., picking up/dropping objects).\n"
            "- StateAction: Changes internal state (e.g., navigation target).\n\n"
            "Rules:\n"
            "- DO NOT negate any condition or node.\n"
            "- ONLY use the following behaviors:\n"
            "{"
            "'drop_part': 'Node Type: ActuatorAction\\nDescription: Makes the agent drop a part only if it\\'s holding one. Returns True if the agent drops a part, and False otherwise.', "
            "'is_agent_holding_good_part': 'Node Type: Condition\\nDescription: Checks whether the agent is holding a good part. Returns True if the agent is holding a good part, and False otherwise.', "
            "'is_agent_holding_scrap_part': 'Node Type: Condition\\nDescription: Checks whether the agent is holding a scrap part. Returns True if the agent is holding a scrap part, and False otherwise.', "
            "'is_agent_in_base_area': 'Node Type: Condition\\nDescription: Checks whether the agent is in the base area.', "
            "'is_agent_in_construction_area': 'Node Type: Condition\\nDescription: Checks whether the agent is in the construction zone within the base area.', "
            "'is_agent_in_source_area': 'Node Type: Condition\\nDescription: Checks whether the agent is in the source area.', "
            "'is_agent_in_storage_area': 'Node Type: Condition\\nDescription: Checks whether the agent is in the storage zone within the base area.', "
            "'is_agent_in_waste_area': 'Node Type: Condition\\nDescription: Checks whether the agent is in the waste area.', "
            "'is_good_part_detected': 'Node Type: Condition\\nDescription: Checks whether the agent detects a good part within range.', "
            "'is_scrap_part_detected': 'Node Type: Condition\\nDescription: Checks whether the agent detects a scrap part within range.', "
            "'pick_up_part': 'Node Type: ActuatorAction\\nDescription: Makes the agent pick up a part if within range and not already holding.', "
            "'state_movement_freeze': 'Node Type: StateAction\\nDescription: Freeze and stop movement.', "
            "'state_random_walk': 'Node Type: StateAction\\nDescription: Move randomly.', "
            "'state_seek_base_area': 'Node Type: StateAction\\nDescription: Move toward base.', "
            "'state_seek_construction_area': 'Node Type: StateAction\\nDescription: Move toward construction zone.', "
            "'state_seek_source_area': 'Node Type: StateAction\\nDescription: Move toward source.', "
            "'state_seek_storage_area': 'Node Type: StateAction\\nDescription: Move toward storage.', "
            "'state_seek_waste_area': 'Node Type: StateAction\\nDescription: Move toward waste area.'"
            "}\n\n"
            "USER REQUEST: Generate behavior tree to \"Find a scrap part. If you find one, pick it up.\" "
            "Output only the XML behavior tree without extra text or explanations of the tree.\n\n"
            "RESPONSE:"
        )
    }]
}]

# Apply template
text = tokenizer.apply_chat_template(
    messages,
    add_generation_prompt=True,
)

# Generate output
outputs = model.generate(
    **tokenizer([text], return_tensors="pt").to("cuda"),
    max_new_tokens=1000,
    temperature=1.0,
    top_p=0.95,
    top_k=64,
)

# Decode output
print(tokenizer.batch_decode(outputs, skip_special_tokens=True)[0])


user
You are one of the top robotics professionals in the world. You specialize in using behavior trees in XML format to control agents and robots. You are working on a project that is harnessing LLMs for the control of agents in specific environments by allowing a user to specify a task in natural language, and the LLM will produce a behavior tree that should solve that task. The project is focused on an environment regarding base maintenance through a swarm of agents.

Environment Description: In the environment, several parts are scattered around and parts have two types: good parts and scrap parts. The environment consists of 3 main areas. The first zone is the base where agents spawn and is separated between two smaller areas, a storage area and a construction area. The second area is the source area; good parts spawn much more frequently the closer you get to the source, and bad parts while they can be found there, they can appear anywhere except the base area. The last area is t

 You can also use a `TextStreamer` for continuous inference - so you can see the generation token by token, instead of waiting the whole time!

In [ ]:
# messages = [{
#     "role": "user",
#     "content": [{"type" : "text", "text" : "Why is the sky blue?",}]
# }]
# text = tokenizer.apply_chat_template(
#     messages,
#     add_generation_prompt = True, # Must add for generation
# )

# from transformers import TextStreamer
# _ = model.generate(
#     **tokenizer([text], return_tensors = "pt").to("cuda"),
#     max_new_tokens = 64, # Increase for longer outputs!
#     # Recommended Gemma-3 settings!
#     temperature = 1.0, top_p = 0.95, top_k = 64,
#     streamer = TextStreamer(tokenizer, skip_prompt = True),
# )

<a name="Save"></a>
### Saving, loading finetuned models
To save the final model as LoRA adapters, either use Huggingface's `push_to_hub` for an online save or `save_pretrained` for a local save.

**[NOTE]** This ONLY saves the LoRA adapters, and not the full model. To save to 16bit or GGUF, scroll down!

In [ ]:
# model.save_pretrained("gemma-3")  # Local saving
# tokenizer.save_pretrained("gemma-3")
# # model.push_to_hub("HF_ACCOUNT/gemma-3", token = "...") # Online saving
# # tokenizer.push_to_hub("HF_ACCOUNT/gemma-3", token = "...") # Online saving

Now if you want to load the LoRA adapters we just saved for inference, set `False` to `True`:

In [ ]:
# if False:
#     from unsloth import FastModel
#     model, tokenizer = FastModel.from_pretrained(
#         model_name = "lora_model", # YOUR MODEL YOU USED FOR TRAINING
#         max_seq_length = 2048,
#         load_in_4bit = True,
#     )

# messages = [{
#     "role": "user",
#     "content": [{"type" : "text", "text" : "What is Gemma-3?",}]
# }]
# text = tokenizer.apply_chat_template(
#     messages,
#     add_generation_prompt = True, # Must add for generation
# )

# from transformers import TextStreamer
# _ = model.generate(
#     **tokenizer([text], return_tensors = "pt").to("cuda"),
#     max_new_tokens = 64, # Increase for longer outputs!
#     # Recommended Gemma-3 settings!
#     temperature = 1.0, top_p = 0.95, top_k = 64,
#     streamer = TextStreamer(tokenizer, skip_prompt = True),
# )

### Saving to float16 for VLLM

We also support saving to `float16` directly for deployment! We save it in the folder `gemma-3-finetune`. Set `if False` to `if True` to let it run!

In [ ]:
# if False: # Change to True to save finetune!
#     model.save_pretrained_merged("gemma-3-finetune", tokenizer)

If you want to upload / push to your Hugging Face account, set `if False` to `if True` and add your Hugging Face token and upload location!

In [ ]:
# if False: # Change to True to upload finetune
#     model.push_to_hub_merged(
#         "HF_ACCOUNT/gemma-3-finetune", tokenizer,
#         token = "hf_..."
#     )

### GGUF / llama.cpp Conversion
To save to `GGUF` / `llama.cpp`, we support it natively now for all models! For now, you can convert easily to `Q8_0, F16 or BF16` precision. `Q4_K_M` for 4bit will come later!

In [ ]:
# if True: # Change to True to save to GGUF
#     model.save_pretrained_gguf(
#         "gemma-3-finetune",
#         quantization_type = "Q4_0", # For now only Q8_0, BF16, F16 supported
#     )

Likewise, if you want to instead push to GGUF to your Hugging Face account, set `if False` to `if True` and add your Hugging Face token and upload location!

In [ ]:
if False: # Change to True to upload GGUF
    model.push_to_hub_gguf(
        "gemma-3-finetune",
        quantization_type = "Q8_0", # Only Q8_0, BF16, F16 supported
        repo_id = "HF_ACCOUNT/gemma-finetune-gguf",
        token = "hf_...",
    )

Now, use the `gemma-3-finetune.gguf` file or `gemma-3-finetune-Q4_K_M.gguf` file in llama.cpp or a UI based system like Jan or Open WebUI. You can install Jan [here](https://github.com/janhq/jan) and Open WebUI [here](https://github.com/open-webui/open-webui)

And we're done! If you have any questions on Unsloth, we have a [Discord](https://discord.gg/unsloth) channel! If you find any bugs or want to keep updated with the latest LLM stuff, or need help, join projects etc, feel free to join our Discord!

Some other links:
1. Train your own reasoning model - Llama GRPO notebook [Free Colab](https://colab.research.google.com/github/unslothai/notebooks/blob/main/nb/Llama3.1_(8B)-GRPO.ipynb)
2. Saving finetunes to Ollama. [Free notebook](https://colab.research.google.com/github/unslothai/notebooks/blob/main/nb/Llama3_(8B)-Ollama.ipynb)
3. Llama 3.2 Vision finetuning - Radiography use case. [Free Colab](https://colab.research.google.com/github/unslothai/notebooks/blob/main/nb/Llama3.2_(11B)-Vision.ipynb)
6. See notebooks for DPO, ORPO, Continued pretraining, conversational finetuning and more on our [documentation](https://docs.unsloth.ai/get-started/unsloth-notebooks)!

<div class="align-center">
  <a href="https://unsloth.ai"><img src="https://github.com/unslothai/unsloth/raw/main/images/unsloth%20new%20logo.png" width="115"></a>
  <a href="https://discord.gg/unsloth"><img src="https://github.com/unslothai/unsloth/raw/main/images/Discord.png" width="145"></a>
  <a href="https://docs.unsloth.ai/"><img src="https://github.com/unslothai/unsloth/blob/main/images/documentation%20green%20button.png?raw=true" width="125"></a>

  Join Discord if you need help + ⭐️ <i>Star us on <a href="https://github.com/unslothai/unsloth">Github</a> </i> ⭐️
</div>
